### DISCUSSION TITLE
Uppercase Š,Č,Ž,...

### DISCUSSION THREAD

**Discussion by Silko5** [simolus3/drift#2756] [Q&A]
I have a search implemented with flutter autocomplete field. It searches through specified table in a database using search string as where condition with LIKE comparison. Everything works perfectly except uppercase utf-8 charcaters such as Š,Č,Ž, etc. won't find anything. As I found out this is probably sqlite problem with utf-8 special characters described here: https://sqlite.org/lang_expr.html#like

Is there nothing that can be done in this case?

**Comment by simolus3** [comment]
I don't this this is possible at the moment. In the future, we can likely provide a custom `LIKE` operator with [intl4x](https://github.com/dart-lang/i18n/tree/main/pkgs/intl4x), but at the moment I wouldn't even know how to do this in Dart. If you have a way to compare these in Dart, I'm happy to help bring that functionality to a custom sqlite3 function you could use.

**Reply by Silko5** [reply]
Hi @simolus3. Thank you for quick response. I don't currently have a way but I must add this:
- lowercase š,č,ž,etc. characters work as intended
- if I type uppercase Š,Č,Ž,etc in autocomplete it finds rows including lowercase characters š,č,ž, etc. but not those with uppercase
- if I type lowercase š,č,ž, etc in autocomplete it also finds rows including lowercase charcters š,č,ž etc. but not those with uppercase
- if I type any other ascii character lower or uppercase it finds rows with lower or uppercase characters

So the problem is only that I can't find rows including uppercase Š,Č,Ž,etc. characters. Maybe with this additional info you have some solution?

EDIT:
Also, I tried hardcoding unicode for uppercase letter Š (which is "\u0160") into select query. And it found rows including uppercase letter Š.

**Comment by Silko5** [comment]
I managed to create a bit of a hacky fix. Maybe you can use it to create custom LIKE function. This is how I build search expression:

```
if (searchText != null) {
  if (searchColumns != null) {
    for (final search in searchColumns!) {
      final column = widget.table!.columnsByName[search];
      if (column != null && column is GeneratedColumn<String>) {
        final textColumn = column;
        final searchExpression = textColumn.like("%$searchText%");
        if (searchExp == null) {
          searchExp = searchExpression;
        } else {
          searchExp = searchExp | searchExpression;
        }

        searchExp = utf8UpperCaseSearchFix(searchText, searchExp, textColumn);
      }
    }
  }
}
```

and this is a utf8UpperCaseSearchFix function:
```
Expression<bool>? utf8UpperCaseSearchFix(
      String searchText,
      Expression<bool>? searchExp,
      GeneratedColumn<String> textColumn
) {
  String? additionalSearchString;

  Map<String, String> unicodeMap = {
    "š": "\u0160",
    "č": "\u010C",
    "ž": "\u017D",
    "ć": "\u0106",
    "đ": "\u0110",
  };

  unicodeMap.forEach((key, value) {
    if(searchText.contains(key)) {
      if(additionalSearchString != null) {
        additionalSearchString = additionalSearchString!.replaceAll(key, value);
      }
      else {
        additionalSearchString = searchText.replaceAll(key, value);
      }
    }
  });

  if(additionalSearchString != null) {
    final additionalSearchExpression = textColumn.like("%$additionalSearchString%");
    if (searchExp == null) {
      searchExp = additionalSearchExpression;
    } else {
      searchExp = searchExp | additionalSearchExpression;
    }
  }

  return searchExp;
}
```
